### PR TITLE
Parallelisation

### PR DESCRIPTION
**Pull request by FWuellhorst**
As described in https://github.com/RWTH-EBC/AixCaliBuHA/issues/20, changes based on enabling MP in optimization. Related to #42

### CODE REVIEW

File: ebcpy/simulationapi/dymola_api.py [hunk 1]
```diff
@@ -230,10 +231,11 @@ def __init__(self, cd, model_name, packages=None, **kwargs):
...
         if not self.debug:
             atexit.register(self.close)
-        if self.use_mp:
-            self.pool.map(self._setup_dymola_interface, [True for _ in range(self.n_cpu)])
         # For translation etc. always setup a default dymola instance
         self.dymola = self._setup_dymola_interface(use_mp=False)
+        if self.use_mp:
+            self.pool.map(self._setup_dymola_interface, [True for _ in range(self.n_cpu)])
+
         self.fully_initialized = True
```
Comment (FWuellhorst) [inline ebcpy/simulationapi/dymola_api.py:239]: Is there a specific reason to change the order? Should work both ways I guess.

File: ebcpy/optimization.py [hunk 2]
```diff
@@ -350,9 +373,30 @@ def _evaluate(self, x, out, *args, **kwargs):
...
             copy_termination = default_kwargs.pop("copy_termination")
 
+            # Get kwargs for algorithm
+            pop_size = kwargs["pop_size"]
+            # GA:
+            sampling = get_sampling(name=kwargs["sampling"])
+            selection = get_selection(name=kwargs["selection"])
+            crossover = get_crossover(name=kwargs["crossover"])
+            mutation = get_mutation(name=kwargs["mutation"])
+            eliminate_duplicates = kwargs["eliminate_duplicates"]
```
Comment (FWuellhorst) [inline ebcpy/optimization.py:383]: Is it necessary to call `get_sampling` etc.? I though pymoo deals with the config?!

**Review by FWuellhorst — COMMENTED** [review]
I got some open questions

**Comment by jonasklingebiel1** [comment]
Hi Fabian,

looks good to me! I changed the pop_size of pymoo optimizer (specified in e4) to 20. Before pop_size and max_iter was both 100 so the GA could not converge. The results for both examples are quite good now. 

The only issue I can report is that when closing fmu (after succesful validation) an error 

```
"Closing fmu for worker 1
Could not terminate fmu instance: fmi2Terminate failed with status 3 (error)."
``` 
is reported. I ran some further tests and came up with a few usability enhancements, some of them should be quite easy to implement:

- [ ] Is there an option to store the data of objective_plot and tuner_parameter_plot as  CSV? This would be quite helpful. I know that the data is stored in the log-file but writing a parsing script that extracts the relevant data is time-consuming and leaves room for error.
- [ ] The automatically updating plots during optimization are very helpful. All three plots are generated on the same window position, so the plots overlap and need to be moved manually. Plotting into one figure using subplots or specifying different plot positions would be helpful
- [ ] After a successful run, writing the directory where the results are stored into the log would be helpful. Further, it would be convenient if an explorer window linking to the result directory is opened automatically.

**Comment by FWuellhorst** [comment]
@jonasklingebiel1 Thanks for the nice ideas! You can open an issue in https://github.com/RWTH-EBC/AixCaliBuHA or we can try to integrate these features before the PR in AixCaliBuHA. To check if AixCaliBuHA works fine, this PR has to come first however. 

Regarding the fmu-error: I don't receive it on single core. Maybe there is something going wrong in multiprocessing. Anyway, on closing, this is not a big issue as it does not affect the simulation quality.

**Review by jonasklingebiel1 — APPROVED** [review]

**Pull request merged**